fdbserver/OldTLogServer_6_0.actor.cpp

```diff
@@ -882,28 +898,69 @@ std::deque<std::pair<Version, LengthPrefixedStringRef>> & getVersionMessages( Re
-	auto tagData = logData->getTagData(req.tag);
+ACTOR Future<Void> tLogPopCore( TLogData* self, Tag tag, Version to, Reference<LogData> logData ) {
+	if (self->ignorePopRequest && tag != txsTag) {
+		TraceEvent("IgnoringPopRequest").detail("IgnorePopDeadline", self->ignorePopDeadline);
+
+		if (self->toBePopped.find(tag) == self->toBePopped.end()
+			|| to > self->toBePopped[tag]) {
+			self->toBePopped[tag] = to;
+		}
+		// add the pop to the toBePopped map
+		TraceEvent(SevDebug, "IgnoringPopRequest")
+			.detail("IgnorePopDeadline", self->ignorePopDeadline)
+			.detail("Tag", tag.toString())
+			.detail("Version", to);
+		return Void();
+	}
+	auto tagData = logData->getTagData(tag);
 	if (!tagData) {
-		tagData = logData->createTagData(req.tag, req.to, true, true, false);
-	} else if (req.to > tagData->popped) {
-		tagData->popped = req.to;
+		tagData = logData->createTagData(tag, to, true, true, false);
+	} else if (to > tagData->popped) {
+		tagData->popped = to;
 		tagData->poppedRecently = true;
 
-		if(tagData->unpoppedRecovered && req.to > logData->recoveredAt) {
+		if(tagData->unpoppedRecovered && to > logData->recoveredAt) {
 			tagData->unpoppedRecovered = false;
 			logData->unpoppedRecoveredTags--;
-			TraceEvent("TLogPoppedTag", logData->logId).detail("Tags", logData->unpoppedRecoveredTags).detail("Tag", req.tag.toString()).detail("DurableKCVer", logData->durableKnownCommittedVersion).detail("RecoveredAt", logData->recoveredAt);
+			TraceEvent("TLogPoppedTag", logData->logId).detail("Tags", logData->unpoppedRecoveredTags).detail("Tag", tag.toString()).detail("DurableKCVer", logData->durableKnownCommittedVersion).detail("RecoveredAt", logData->recoveredAt);
 			if(logData->unpoppedRecoveredTags == 0 && logData->durableKnownCommittedVersion >= logData->recoveredAt && logData->recoveryComplete.canBeSet()) {
 				logData->recoveryComplete.send(Void());
 			}
 		}
 
-		if ( req.to > logData->persistentDataDurableVersion )
-			wait(tagData->eraseMessagesBefore( req.to, self, logData, TaskTLogPop ));
-		//TraceEvent("TLogPop", self->dbgid).detail("Tag", req.tag).detail("To", req.to);
+		if ( to > logData->persistentDataDurableVersion )
+			wait(tagData->eraseMessagesBefore( to, self, logData, TaskTLogPop ));
+		//TraceEvent("TLogPop", self->dbgid).detail("Tag", tag).detail("To", to);
 	}
+	return Void();
+}
 
+ACTOR Future<Void> tLogPop( TLogData* self, TLogPopRequest req, Reference<LogData> logData ) {
+	// timeout check for ignorePopRequest
+	if (self->ignorePopRequest && (g_network->now() > self->ignorePopDeadline)) {
+
+		TraceEvent("EnableTLogPlayAllIgnoredPops");
+		// use toBePopped and issue all the pops
+		state std::map<Tag, Version>::iterator it;
+		for (it = self->toBePopped.begin(); it != self->toBePopped.end(); it++) {
+			TraceEvent("PlayIgnoredPop")
+				.detail("Tag", it->first.toString())
+				.detail("Version", it->second);
+			wait(tLogPopCore(self, it->first, it->second, logData));
```

Comment:
I'd think you'd want to do these in parallel?

Reply:
Since tLogPopCore() is yielding, it may be useful to spawn all the tLogPopCore simultaneously.